src/contexts/PortalTreeContext.ts

```diff
@@ -0,0 +1,22 @@
+import { ReactPortal, createContext } from "react";
+
+/**
+ * The "Portal Tree" is a registry of React Portals, indexed by
+ * the "key" of their parent node views.
+ *
+ * Each node view registers a portal with the tree under its
+ * parent's key. Parents then retrieve the list of child portals
+ * under their key, allowing portals to be rendered with the
+ * appropriate hierarchy.
+ */
+export const PortalTreeContext = createContext<Record<string, ReactPortal[]>>(
+  null as unknown as Record<string, ReactPortal[]>
+);
+
+/**
+ * Node views that don't have any React node view ancestors
+ * can specify their parent node view as the "root", and
+ * will be have their portals rendered as direct children of
+ * the ProseMirror component.
+ */
+export const PORTAL_TREE_ROOT_KEY = "root";
```

Comment:
Would this be an appropriate use of `Symbol`?

Reply:
Yeah, seems like it!

Reply:
I guess if this is going to be put in a DOM node's dataset it might need to be a string.

Reply:
It's not, though! This is the fallback for when we couldn't find a DOM node ancestor

Reply:
Ah, yes. Cool.